java/configuration/src/main/java/sleeper/configuration/table/index/DynamoDBTableIndex.java

```diff
@@ -135,13 +136,82 @@ public void delete(TableId tableId) {
+            CancellationReason idNotFoundReason = e.getCancellationReasons().get(1);
+            if (isCheckFailed(nameNotFoundReason)) {
+                throw TableNotFoundException.withTableName(tableId.getTableName());
+            } else if (isCheckFailed(idNotFoundReason)) {
+                throw TableNotFoundException.withTableId(tableId.getTableUniqueId());
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    @Override
+    public void update(TableId tableId) {
+        TableId oldId = getTableByUniqueId(tableId.getTableUniqueId())
+                .orElseThrow(() -> TableNotFoundException.withTableId(tableId.getTableUniqueId()));
+        update(oldId, tableId);
+    }
+
+    public void update(TableId oldId, TableId newId) {
+        Map<String, AttributeValue> idItem = DynamoDBTableIdFormat.getItem(newId);
+        TransactWriteItemsRequest request = new TransactWriteItemsRequest()
+                .withReturnConsumedCapacity(ReturnConsumedCapacity.TOTAL)
+                .withTransactItems(
+                        new TransactWriteItem().withPut(new Put()
+                                .withTableName(nameIndexDynamoTableName)
+                                .withItem(idItem)
+                                .withConditionExpression("attribute_not_exists(#tablename)")
+                                .withExpressionAttributeNames(Map.of("#tablename", TABLE_NAME_FIELD))),
+                        new TransactWriteItem().withPut(new Put()
+                                .withTableName(idIndexDynamoTableName)
+                                .withItem(idItem)
+                                .withConditionExpression("attribute_exists(#tableid)")
+                                .withExpressionAttributeNames(Map.of("#tableid", TABLE_ID_FIELD))),
+                        new TransactWriteItem().withDelete(new Delete()
+                                .withTableName(nameIndexDynamoTableName)
+                                .withKey(DynamoDBTableIdFormat.getNameKey(oldId))
+                                .withConditionExpression("attribute_exists(#tablename)")
+                                .withExpressionAttributeNames(Map.of("#tablename", TABLE_NAME_FIELD))));
+        try {
+            TransactWriteItemsResult result = dynamoDB.transactWriteItems(request);
+            List<ConsumedCapacity> consumedCapacity = result.getConsumedCapacity();
+            double totalCapacity = consumedCapacity.stream().mapToDouble(ConsumedCapacity::getCapacityUnits).sum();
+            LOGGER.debug("Updated table {}, capacity consumed = {}", newId, totalCapacity);
+        } catch (TransactionCanceledException e) {
+            CancellationReason nameAlreadyExistsReason = e.getCancellationReasons().get(0);
+            CancellationReason idNotFoundReason = e.getCancellationReasons().get(1);
+            CancellationReason oldNameNotFoundReason = e.getCancellationReasons().get(2);
```

Comment:
These look fragile. Is it guaranteed that `e.getCancellationReasons` will always have 3 reasons in the above order?

Reply:
When a transaction is canceled you get a CancellationReason for each transaction item, see the notes under TransactionCanceledException in the docs:

https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_TransactWriteItems.html#API_TransactWriteItems_Errors